### PR TITLE
[WIP] CORTX-30726: Generate Machine-id for all Nodes from FQDN/any string coming from solution(cluster.yaml)

### PR DESCRIPTION
**Pull request by tanujashinde0405**
# Problem Statement
Objectives

Anything that provisioner does must be transparent to the component, i.e. components should not have to change anything. 
ID treatment must be same for all Pods / containers. No special treatment for Server or Data Pods. 
Minimise the amount of business logic around machine-id in k8s code. 
How this can be done. 

Provisioner will generate /etc/cortx/machine-id
CORTX image will have link from /etc/machine-id -> /etc/cortx/machine-id

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [x] Side effects on other features (deployment/upgrade)
- [x] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: src/provisioner/cluster.py [hunk 1]
```diff
@@ -75,6 +77,12 @@ def add_node(self, node: dict):
         self._validate(node)
         self._node_list.append(node)
 
+    @staticmethod
+    def get_machine_id(machine_id: str):
```
Comment (ujjwalpl) [inline src/provisioner/cluster.py:81]: input name should be node_id to avoid confusion.

File: src/provisioner/cluster.py [hunk 2]
```diff
@@ -75,6 +77,12 @@ def add_node(self, node: dict):
...
+        result = hashlib.md5(machine_id.encode())
+        node_id= result.hexdigest()
+        return node_id
```
Comment (ujjwalpl) [inline src/provisioner/cluster.py:84]: Return should be machine_id as the function name suggests.

File: src/provisioner/cluster.py [hunk 3]
```diff
@@ -124,17 +132,21 @@ def _get_storage_kv_list(self, storage_spec: dict, node_id: str):
 
     def save(self, cortx_conf):
         """ Saves cluster information onto the conf store """
-
         kvs = []
         try:
             for node in self._node_list:
-                node_id = node.pop('id')
+                logical_id = node.pop('id')
```
Comment (ujjwalpl) [inline src/provisioner/cluster.py:138]:
```
node_id = node.pop('id')
machine_id = CortxCluster.get_machine_id(node_id)

File: src/provisioner/cluster.py [hunk 4]
```diff
@@ -124,17 +132,21 @@ def _get_storage_kv_list(self, storage_spec: dict, node_id: str):
...
+                node_id = CortxCluster.get_machine_id(logical_id)
                 key_prefix = f'node>{node_id}'
+                if socket.gethostname() == node['hostname'] :
```
Comment (ujjwalpl) [inline src/provisioner/cluster.py:141]: Remove this condition. Logic should be common for all pods unconditionally.

File: src/provisioner/cluster.py [hunk 5]
```diff
@@ -201,7 +213,8 @@ def save(self, cortx_conf):
                 nodes = storage_set['nodes']
                 for node_idx, node in enumerate(nodes):
                     # confstore keys
-                    kvs.append((f'{key_prefix}>nodes[{node_idx}]', node['id']))
+                    node_id = CortxCluster.get_machine_id(node['id'])
+                    kvs.append((f'{key_prefix}>nodes[{node_idx}]', node_id))
```
Comment (ujjwalpl) [inline src/provisioner/cluster.py:217]: I think machine_id needs to reflect in GConf and not to the solution supplied config. 
Key would be "storage_sets[i]>nodes[j]>id".

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/provisioner/cluster.py (hunk 4 above)
Comment (tanujashinde0405) [reply src/provisioner/cluster.py:141]: As discussed in the call, we have to keep the if condition as of now to recognize the current pod with hostname.

File: src/provisioner/cluster.py (hunk 5 above)
Comment (tanujashinde0405) [reply src/provisioner/cluster.py]: done, I have replcaed it with machine_id now.

File: src/provisioner/cluster.py (hunk 2 above)
Comment (tanujashinde0405) [reply src/provisioner/cluster.py]: done

File: src/provisioner/cluster.py (hunk 1 above)
Comment (tanujashinde0405) [reply src/provisioner/cluster.py]: done

**Comment by kailashyadav11** [comment]
Please run the unit tests related to  the change and add new testcases if needed.

**Review by nitin-seagate — APPROVED** [review]
LGTM

**Review by ujjwalpl — APPROVED** [review]

**Comment by tanujashinde0405** [comment]
> Please run the unit tests related to the change and add new test cases if needed.

Ran the tests, and everything is working, no new test cases are required. Samples files under the test folder were outdated, I have updated them as well as updated readme.md.

**Review by kailashyadav11 — APPROVED** [review]

File: src/provisioner/cluster.py [hunk 6]
```diff
@@ -124,17 +132,21 @@ def _get_storage_kv_list(self, storage_spec: dict, node_id: str):
...
-
         kvs = []
         try:
             for node in self._node_list:
                 node_id = node.pop('id')
-                key_prefix = f'node>{node_id}'
+                machine_id = CortxCluster.get_machine_id(node_id)
+                key_prefix = f'node>{machine_id}'
+                if socket.gethostname() == node['hostname'] :
+                    machine_id_path = open(const.MACHINE_ID_PATH, 'w')
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply src/provisioner/cluster.py]: done

File: src/provisioner/cluster.py [hunk 7]
```diff
@@ -124,17 +132,21 @@ def _get_storage_kv_list(self, storage_spec: dict, node_id: str):
...
-
         kvs = []
         try:
             for node in self._node_list:
-                node_id = node.pop('id')
+                logical_id = node.pop('id')
+                node_id = CortxCluster.get_machine_id(logical_id)
                 key_prefix = f'node>{node_id}'
+                if socket.gethostname() == node['hostname'] :
+                    machine_id = open(const.MACHINE_ID_PATH, 'w')
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply src/provisioner/cluster.py]: done

**Pull request closed without merging**